qiskit/result/counts.py

```diff
@@ -103,18 +103,38 @@ def __init__(self, data, time_taken=None, creg_sizes=None, memory_slots=None):
                     "Invalid input key type %s, must be either an int "
                     "key or string key with hexademical value or bit string"
                 )
-        header = {}
+
+        if not bin_data:
+            bin_data = postprocess.format_counts(
+                self.hex_raw,
+                header={"creg_sizes": creg_sizes, "memory_slots": memory_slots},
+            )
+
+        super().__init__({k: v for k, v in bin_data.items() if v != 0})
+
+        if not memory_slots:
+            if creg_sizes:
+                memory_slots = sum((i[1] for i in creg_sizes))
+            else:
+                memory_slots = max([len(k) for k in bin_data.keys()]) if bin_data else 0
+
         self.creg_sizes = creg_sizes
-        if self.creg_sizes:
-            header["creg_sizes"] = self.creg_sizes
         self.memory_slots = memory_slots
-        if self.memory_slots:
-            header["memory_slots"] = self.memory_slots
-        if not bin_data:
-            bin_data = postprocess.format_counts(self.hex_raw, header=header)
-        super().__init__(bin_data)
         self.time_taken = time_taken
 
+    def __getitem__(self, key):
+        try:
+            return super().__getitem__(key)
+        except KeyError as keyerror:
+            if isinstance(key, int):
+                key = hex(key)
+            key = key.replace(" ", "")
+            if not self.bitstring_regex.search(key):
+                key = postprocess.format_counts_memory(key, {"memory_slots": self.memory_slots})
+            if self.memory_slots == len(key):
+                return 0
+            raise KeyError from keyerror
```

Comment:
So `__getitem__` was in the comment at the top for a reason, updating dunder methods for a dict subclass is not a good idea and  is explicitly what we should not be doing. Specifically here `get()` will  not have this extra logic now. Basically if you do `get('011', 24)` where `011` is not in the list it will return 24 wihle if you do counts['011'] will return 0. You don't want this to be mismatched like that because it will cause confusing bugs down the road.

The only way to make this work reliably is to change `Counts` to be a `Mapping` implementation and not a dict subclass.

Reply:
done in 92f02a3c8 . Thanks!